torch/csrc/api/include/torch/nn/utils/clip_grad.h

```diff
@@ -26,7 +27,7 @@ inline double clip_grad_norm_(
   if (norm_type == std::numeric_limits<double>::infinity()) {
     for (const auto& param : params_with_grad) {
       auto param_max = param.grad().data().abs().max().item().toDouble();
-      if (param_max > total_norm) {
+      if (param_max > total_norm || std::isnan(param_max)) {
```

Comment:
Note: I found out this was not accumulating nan values, so had to add this check

Reply:
I think this causes the C++ impl to diverge from python's. On the python side, the logic is:
```python
...
if norm_type == inf:
    total_norm = max(p.grad.detach().abs().max().to(device) for p in parameters)
...
```
Note that `max()` seems to ignore NaNs:
```python
>>> max([1., 2., 1000, 1e20, float('nan')])
1e+20
```

Reply:
Hmm, this is a bit odd:

```python
>>> max([float('nan'), 0, 1, 3])
nan
>>> max([3, float('nan'), 0, 1])
3
```

If NaN is the first value given to `max()`, it gets returned. Otherwise, NaNs are ignored. This seems to come from the fact that all comparisons with NaN are false.

There are a couple problems to solve here:
* Before this PR, the C++ impl did not match the python impl's handling of NaNs. If the Python impl's `max()` call starts with a NaN, it would return NaN. But if the C++ impl was given the same input, it would ignore the NaN completely. So if the `max()` call was given a 1-element list with one NaN, the python impl returns NaN and the C++ impl returns 0.
* In this PR, we are currently checking whether the `total_norm` (the norm of all the inputs' norms--or in the case of `norm_type=inf`, the max of all the inputs` maxes) is NaN or inf. For this check to be correct, we would need proper NaN propagation. Otherwise, we would have to check the inputs explicitly, which is time consuming.

There are a couple options to resolve this, I think
1. Fix both the Python and C++ impls to propagate NaNs. This would probably mean using `torch.max()`. I realize this might be BC breaking, but I think it makes clip_grad_norm_ more consistent with itself, because if `norm_type` is not `inf`, then we get proper NaN propagation through both `torch.norm()` calls.
2. Change the C++ impl to match Python. Both impls would have NaN propagation only if the first tensor given to clip_grad_norm_ has a NaN in its grad. I don't really like this quirky behavior. Furthermore, since NaNs don't get propagated properly, we would have to check each input for NaNs.
3. Change the Python impl to match C++. Since this would mean there would be absolutely no NaN propagation for `norm_type=inf`, we'd have to check each input.

So far, I only like solution 1. Does anyone have a different perspective?

Reply:
> If NaN is the first value given to max(), it gets returned. Otherwise, NaNs are ignored. This seems to come from the fact that all comparisons with NaN are false.

Great find! I agree that this quirky python behavior is undesirable so I prefer solution 1 as well, but as this will result in more restrictive behavior than before, I want to be careful making this BC-breaking change.

`torch.norm()` propagates NaNs correctly for the infinity norm:
```python
>>> torch.norm(torch.tensor([1., 2., float('nan')]), p=float('inf'))
tensor(nan)
>>> torch.norm(torch.tensor([float('nan'), 1., 2.]), p=float('inf'))
tensor(nan)
>>> torch.norm(torch.tensor([1., float('nan'), 2.]), p=float('inf'))
tensor(nan)
```

To me, this begs the question: why did the original implementation explicitly re-implement the norm calculation with `max()` instead of using `torch.norm()` to calculate the infinity norm?

I'm going to do a bit of investigation on this before deciding on the direction to go.

Reply:
Yeah, `torch.norm()` probably would be a good way to go. Maybe it didn't have infinite order support back when `clip_grad_norm_` was first written

Reply:
Yeah, most likely reasons are historic, and I agree with your solution 1)  @kurtamohler, we should correctly propagate nans.

Reply:
Update: the [original commit](https://github.com/pytorch/pytorch/commit/c2c1710047983e18560bb9736261620b9ac4b58f) calculated all norms manually since `torch.norm()` didn't exist back then.

[This commit](https://github.com/pytorch/pytorch/commit/e74a215adeec1791cff85376f056c6e04828bcfb#diff-f89b14a0391acb90e2ee7930f942032f0376b64a251066d6760701e725fa0ee7) changed non-inf norm calculation to use `torch.norm()` (removing the calls to `item()`) for performance.

I wonder how performance is when using `torch.max()` or `torch.norm()` compared to the current implementation?

Reply:
My gut impression is that `torch.max()` or `torch.norm()` have the potential to be either faster than python `max()` or no significant difference. Although, there's a chance that would depend on the number of `parameters` being significantly lower than the number of elements in each gradient, which I'd guess is the most common case anyway.

I could do some benchmarking to make sure.

Reply:
I looked into what the performance of `clip_grad_norm_` would be if we replace the python `max` call with `torch.max`. This is the original line from the python impl:
```python
total_norm = max(p.grad.detach().abs().max().to(device) for p in parameters)
```

I figured there were at least two possible ways to replace this line to use `torch.max`

1. Use `torch.stack` to combine all the individual max outputs into one tensor, and then call `torch.max`. Like this:
   ```python
   norms = [p.grad.detach().abs().max().to(device) for p in parameters]
   if len(norms) == 1:
       total_norm = norms[0]
   else:
       total_norm = torch.max(torch.stack(norms))
   ```
2. Preallocate a tensor to store all the individual max outputs, and then call `torch.max` on it. Need to be careful to create the correct dtype tensor, so have to iterate through the parameters and call `torch.promote_types` on each of them. Like this:
   ```python
   if len(parameters) == 1:
       total_norm = parameters[0].grad.detach().abs().max()
   else:   
       dtype = parameters[0].dtype
       for p in parameters[1:]:
           dtype = torch.promote_types(dtype, p.dtype)
       norms = torch.empty(len(parameters), device=device, dtype=dtype)
       for p_idx, p in enumerate(parameters):
           norms[p_idx] = p.grad.detach().abs().max()
       total_norm = torch.max(norms)
   ```

Note, I special-cased `len(parameters) == 1` for both of these because without it, that case had significant overhead costs.

I've checked that both of the above implementations the `clip_grad_norm` tests in `tests_nn.py`

I measured performance with the scripts here: https://github.com/kurtamohler/pytorch-perf-test-scripts/tree/master/clip_grad_norm

Here are my results: https://docs.google.com/spreadsheets/d/1nKTzn2kzrsg6DrYPW54cA8AFQ_3dEhKIh8zDlFQgO48/edit?usp=sharing

Based on this this information, I think the stacked implementation is suitable. In most cases, it matches or exceeds the original performance. In some cases, it's a slowdown, but I think the benefits outweigh the costs--especially considering that we'll get proper NaN propagation from it.

The preallocated implementation seems to have OK performance as well, but not as fast as the stacked implementation in most cases, and the code is a little more complicated, so not desirable.

Reply:
If there are any ideas for an even better implementation, I'm happy to try them out. In the meantime, I'll go ahead with the `torch.stack` impl for now

Reply:
Awesome, thanks for looking into this! I like the stack implementation as well. Happy to get proper NaN propagation with similar performance.